# Week3 7월 16일 월요일 

# Scraping - HTTP

In [2]:
!pip install builtwith

  Running setup.py bdist_wheel for builtwith: started
  Running setup.py bdist_wheel for builtwith: finished with status 'done'
  Stored in directory: C:\Users\soug9\AppData\Local\pip\Cache\wheels\2b\00\c2\a96241e7fe520e75093898bf926764a924873e0304f10b2524
Successfully built builtwith


distributed 1.21.8 requires msgpack, which is not installed.


* builtwith : 기술 정보(웹 서버)를 알려줌

In [1]:
import builtwith

In [4]:
builtwith.parse("http://www.google.com")

{'web-servers': ['Google Web Server']}

In [5]:
builtwith.parse("http://www.naver.com")

{}

In [7]:
builtwith.parse("http://www.wordpress.com")

{'web-servers': ['Nginx'],
 'font-scripts': ['Google Font API'],
 'ecommerce': ['WooCommerce'],
 'cms': ['WordPress'],
 'programming-languages': ['PHP'],
 'blogs': ['PHP', 'WordPress']}

In [8]:
builtwith.parse("http://www.daum.net")

{}

In [22]:
!pip install python-whois

  Running setup.py bdist_wheel for python-whois: started
  Running setup.py bdist_wheel for python-whois: finished with status 'done'
  Stored in directory: C:\Users\soug9\AppData\Local\pip\Cache\wheels\06\cb\7d\33704632b0e1bb64460dc2b4dcc81ab212a3d5e52ab32dc531
  Running setup.py bdist_wheel for future: started
  Running setup.py bdist_wheel for future: finished with status 'done'
  Stored in directory: C:\Users\soug9\AppData\Local\pip\Cache\wheels\bf\c9\a3\c538d90ef17cf7823fa51fc701a7a7a910a80f6a405bf15b1a
Successfully built python-whois future


* 웹사이트 주인 정보

In [23]:
import whois

In [24]:
whois.whois("naver.com") 

{'domain_name': ['NAVER.COM', 'naver.com'],
 'registrar': 'Gabia, Inc.',
 'whois_server': 'whois.gabia.com',
 'referral_url': None,
 'updated_date': [datetime.datetime(2016, 8, 5, 6, 37, 57),
  datetime.datetime(2018, 2, 28, 11, 27, 15)],
 'creation_date': [datetime.datetime(1997, 9, 12, 4, 0),
  datetime.datetime(1997, 9, 12, 0, 0)],
 'expiration_date': [datetime.datetime(2023, 9, 11, 4, 0),
  datetime.datetime(2023, 9, 11, 0, 0)],
 'name_servers': ['NS1.NAVER.COM',
  'NS2.NAVER.COM',
  'ns1.naver.com',
  'ns2.naver.com'],
 'status': ['clientDeleteProhibited https://icann.org/epp#clientDeleteProhibited',
  'clientTransferProhibited https://icann.org/epp#clientTransferProhibited',
  'clientUpdateProhibited https://icann.org/epp#clientUpdateProhibited',
  'ok https://icann.org/epp#ok'],
 'emails': ['white.4818@navercorp.com',
  'dl_ssl@navercorp.com',
  'abuse@gabia.com'],
 'dnssec': 'unsigned',
 'name': 'NAVER Corp.',
 'org': 'NAVER Corp.',
 'address': '6 Buljung-ro, Bundang-gu, Seongn

* 웹 사이트 다운로드 urllib

In [2]:
from urllib.request import urlopen

In [2]:
req = urlopen("http://python.org")

In [3]:
type(req.read())

bytes

In [8]:
req.read()

b''

In [5]:
text = req.read().decode("utf-8")

In [6]:
text

''

In [10]:
urlopen("https://www.google.co.kr/search?q=%ED%95%9C%EA%B8%80&oq=%ED%95%9C%EA%B8%80&aqs=chrome..69i57j69i59j69i61l2.5689j0j7&sourceid=chrome&ie=UTF-8")

HTTPError: HTTP Error 403: Forbidden

In [11]:
urlopen("https://search.naver.com/search.naver?where=nexearch&sm=top_hty&fbm=1&ie=utf8&query=%ED%95%9C%EA%B5%AD")

HTTPError: HTTP Error 403: Forbidden

search는 허용을 하지 않으므로 서버가 막아서 에러가 남 -> error 403

header에 user-agent 정보가 없으므로 서버가 정상적인 사용자가 아니라고 판단(봇이라고 판단)하고 접속을 막음

In [3]:
from urllib.error import HTTPError

In [12]:
try :
    urlopen("https://search.naver.com/search.naver?where=nexearch&sm=top_hty&fbm=1&ie=utf8&query=%ED%95%9C%EA%B5%AD")
except HTTPError as e:
    print(e.code, e.reason, e.headers)

403 Forbidden Date: Mon, 16 Jul 2018 04:27:29 GMT
Content-Type: text/html; charset=UTF-8
Transfer-Encoding: chunked
Connection: close
Vary: Accept-Encoding
Set-Cookie: page_uid=T0wPVdpySDwssvtMZBwssssssvl-469050; path=/; domain=.naver.com
Set-Cookie: _naver_usersession_=99fvmBbLb0fECuXIv9XO2g==; path=/; expires=Mon, 16-Jul-18 04:32:29 GMT; domain=.naver.com
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; report=/p/er/post/xss
Cache-Control: no-cache, no-store, must-revalidate, max-age=0
Pragma: no-cache
Vary: Accept-Encoding




header에 user-agent에 대한 정보가 없어서 에러가 남

1) urllib.request.Request

In [4]:
from urllib.request import Request

In [5]:
# 500대 에러(서버 에러)가 나면 num_retries 만큼 반복해서 urlopen을 시도하는 함수

def download(url, agent="python bot", num_retries=2) :
    headers = {"user-agent" : agent} # header에 user-agent 가 없어서 만들어줌
    req = Request(url, headers = headers) # 입력된 헤더를 가지고 요청함
    
    try :
        resp = urlopen(req)
        
    except HTTPError as e:
        resp = None
        
        print(e.code, e.reason, e.headers)
        
        if 500 <= e.code <=600 and num_retries != 0:
            return download(url, num_retries = num_retries-1)
        
    return resp

In [6]:
download("http://httpstat.us/500")

500 Internal Server Error Cache-Control: private
Server: Microsoft-IIS/10.0
X-AspNetMvc-Version: 5.1
Access-Control-Allow-Origin: *
X-AspNet-Version: 4.0.30319
X-Powered-By: ASP.NET
Set-Cookie: ARRAffinity=8d9742d8920fc8f87dbfb6e900a1f176e365d58296d2c5e44961e07bf1b65817;Path=/;HttpOnly;Domain=httpstat.us
Date: Mon, 16 Jul 2018 07:27:02 GMT
Connection: close
Content-Length: 0


500 Internal Server Error Cache-Control: private
Server: Microsoft-IIS/10.0
X-AspNetMvc-Version: 5.1
Access-Control-Allow-Origin: *
X-AspNet-Version: 4.0.30319
X-Powered-By: ASP.NET
Set-Cookie: ARRAffinity=8d9742d8920fc8f87dbfb6e900a1f176e365d58296d2c5e44961e07bf1b65817;Path=/;HttpOnly;Domain=httpstat.us
Date: Mon, 16 Jul 2018 07:27:04 GMT
Connection: close
Content-Length: 0


500 Internal Server Error Cache-Control: private
Server: Microsoft-IIS/10.0
X-AspNetMvc-Version: 5.1
Access-Control-Allow-Origin: *
X-AspNet-Version: 4.0.30319
X-Powered-By: ASP.NET
Set-Cookie: ARRAffinity=8d9742d8920fc8f87dbfb6e900a1f176e3

In [7]:
download("https://search.naver.com/search.naver?where=nexearch&sm=top_hty&fbm=1&ie=utf8&query=%ED%95%9C%EA%B5%AD")
# agent가 python-bot이니깐

403 Forbidden Date: Mon, 16 Jul 2018 07:27:13 GMT
Content-Type: text/html; charset=UTF-8
Transfer-Encoding: chunked
Connection: close
Vary: Accept-Encoding
Set-Cookie: page_uid=T0x40dpySDCssvvX6JZsssssswo-144322; path=/; domain=.naver.com
Set-Cookie: _naver_usersession_=7D0n9MTUoxmYg42SiH03Ew==; path=/; expires=Mon, 16-Jul-18 07:32:13 GMT; domain=.naver.com
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; report=/p/er/post/xss
Cache-Control: no-cache, no-store, must-revalidate, max-age=0
Pragma: no-cache
Vary: Accept-Encoding




In [9]:
agent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.99 Safari/537.36"

agent를 진짜 사람인척 함... 크롬 개발자도구 network에서 헤더 찾아서 똑같이

In [10]:
html = download("https://search.naver.com/search.naver?where=nexearch&sm=top_hty&fbm=1&ie=utf8&query=%ED%95%9C%EA%B5%AD", agent = agent)

In [11]:
# 출력
html.read().decode("utf-8")

'<!doctype html> <html lang="ko"> <head> <meta charset="utf-8"> <meta name="referrer" content="always">  <meta name="format-detection" content="telephone=no,address=no,email=no"> <meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=2.0"> <meta property="og:title" content="한국 : 네이버 통합검색"/> <meta property="og:image" content="https://ssl.pstatic.net/sstatic/search/common/og_v3.png"> <meta property="og:description" content="\'한국\'의 네이버 통합검색 결과입니다."> <meta name="description" lang="ko" content="\'한국\'의 네이버 통합검색 결과입니다."> <title>한국 : 네이버 통합검색</title> <link rel="shortcut icon" href="https://ssl.pstatic.net/sstatic/search/favicon/favicon_140327.ico">  <link rel="search" type="application/opensearchdescription+xml" href="https://ssl.pstatic.net/sstatic/search/opensearch-description.https.xml" title="Naver" /><link rel="stylesheet" type="text/css" href="https://ssl.pstatic.net/sstatic/search/pc/css/search1_180712.css"> <link rel="stylesheet" type="text/css" href="https

2) requests.request - get 방식

In [12]:
import requests

In [13]:
# requests.request - get

def download_rq(url, agent="python bot", num_retries=2) :
    headers = {"user-agent" : agent}
    resp = requests.request('get', url, headers = headers) # urlopen 과정이 필요없음
    
    if 500 <= resp.status_code <=600 and num_retries != 0:
        print(resp.status_code, resp.reason)
        return download(url, num_retries = num_retries-1)

    return resp

In [14]:
html_rq = download_rq("https://search.naver.com/search.naver?where=nexearch&sm=top_hty&fbm=1&ie=utf8&query=%ED%95%9C%EA%B5%AD", agent = agent)

In [54]:
html_rq

<Response [200]>

In [55]:
type(html_rq)

requests.models.Response

In [15]:
#출력 - decoding 과정 필요 없음
html_rq.text

'<!doctype html> <html lang="ko"> <head> <meta charset="utf-8"> <meta name="referrer" content="always">  <meta name="format-detection" content="telephone=no,address=no,email=no"> <meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=2.0"> <meta property="og:title" content="한국 : 네이버 통합검색"/> <meta property="og:image" content="https://ssl.pstatic.net/sstatic/search/common/og_v3.png"> <meta property="og:description" content="\'한국\'의 네이버 통합검색 결과입니다."> <meta name="description" lang="ko" content="\'한국\'의 네이버 통합검색 결과입니다."> <title>한국 : 네이버 통합검색</title> <link rel="shortcut icon" href="https://ssl.pstatic.net/sstatic/search/favicon/favicon_140327.ico">  <link rel="search" type="application/opensearchdescription+xml" href="https://ssl.pstatic.net/sstatic/search/opensearch-description.https.xml" title="Naver" /><link rel="stylesheet" type="text/css" href="https://ssl.pstatic.net/sstatic/search/pc/css/search1_180712.css"> <link rel="stylesheet" type="text/css" href="https

In [16]:
type(html_rq.text)

str

In [17]:
html_rq.encoding

'UTF-8'

* urllib 한글

In [18]:
from urllib import parse

In [19]:
q = parse.quote("한국")

In [65]:
type(q)

str

In [66]:
q # hexa byte 형태

'%ED%95%9C%EA%B5%AD'

https://search.naver.com/search.naver?

where=nexearch

&sm=top_hty

&fbm=1

&ie=utf8

&query=%ED%95%9C%EA%B5%AD <- 이것이다

In [20]:
url = "https://search.naver.com/search.naver?where=nexearch&sm=top_hty&fbm=1&ie=utf8&query="+q

In [21]:
url

'https://search.naver.com/search.naver?where=nexearch&sm=top_hty&fbm=1&ie=utf8&query=%ED%95%9C%EA%B5%AD'

-- urllib.request 사용

In [71]:
html2 = download(url, agent=agent)

In [72]:
html2.read().decode('utf-8')

'<!doctype html> <html lang="ko"> <head> <meta charset="utf-8"> <meta name="referrer" content="always">  <meta name="format-detection" content="telephone=no,address=no,email=no"> <meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=2.0"> <meta property="og:title" content="한국 : 네이버 통합검색"/> <meta property="og:image" content="https://ssl.pstatic.net/sstatic/search/common/og_v3.png"> <meta property="og:description" content="\'한국\'의 네이버 통합검색 결과입니다."> <meta name="description" lang="ko" content="\'한국\'의 네이버 통합검색 결과입니다."> <title>한국 : 네이버 통합검색</title> <link rel="shortcut icon" href="https://ssl.pstatic.net/sstatic/search/favicon/favicon_140327.ico">  <link rel="search" type="application/opensearchdescription+xml" href="https://ssl.pstatic.net/sstatic/search/opensearch-description.https.xml" title="Naver" /><link rel="stylesheet" type="text/css" href="https://ssl.pstatic.net/sstatic/search/pc/css/search1_180712.css"> <link rel="stylesheet" type="text/css" href="https

-- requests.Request 사용

In [73]:
html2_rq = download_rq(url, agent=agent)

In [75]:
html2_rq.text

'<!doctype html> <html lang="ko"> <head> <meta charset="utf-8"> <meta name="referrer" content="always">  <meta name="format-detection" content="telephone=no,address=no,email=no"> <meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=2.0"> <meta property="og:title" content="한국 : 네이버 통합검색"/> <meta property="og:image" content="https://ssl.pstatic.net/sstatic/search/common/og_v3.png"> <meta property="og:description" content="\'한국\'의 네이버 통합검색 결과입니다."> <meta name="description" lang="ko" content="\'한국\'의 네이버 통합검색 결과입니다."> <title>한국 : 네이버 통합검색</title> <link rel="shortcut icon" href="https://ssl.pstatic.net/sstatic/search/favicon/favicon_140327.ico">  <link rel="search" type="application/opensearchdescription+xml" href="https://ssl.pstatic.net/sstatic/search/opensearch-description.https.xml" title="Naver" /><link rel="stylesheet" type="text/css" href="https://ssl.pstatic.net/sstatic/search/pc/css/search1_180712.css"> <link rel="stylesheet" type="text/css" href="https

* request의 parameter method 

1) get

주소 뒤에 그냥 파라미터 붙여서 넘김

편하다

여러가지 데이터 타입을 받을 수 없다

byte 타입만 받음

보안유지 어려움


2) post

주소를 통해 넘기는게 아니라 데이터는 데이터대로, url은 url대로 넘김

데이터가 많을 때

서버에서 작업이 필요한 경우 ex. 회원가입

3) requests.request - post 방식

In [22]:
import json

In [23]:
params = {
    "where" : "nexearch",
    "sm" : "top_hty",
    "fbm" : 1,
    "ie" : "utf8",
    "query" :"%ED%95%9C%EA%B5%AD"
}

In [24]:
jsonParams = json.dumps(params)

In [25]:
jsonParams

'{"where": "nexearch", "sm": "top_hty", "fbm": 1, "ie": "utf8", "query": "%ED%95%9C%EA%B5%AD"}'

In [26]:
# post 방식 - json으로 파라미터 넘김
def download_json(url, agent="python bot", num_retries=2) :
    headers = {"user-agent" : agent} 
    resp = requests.request('post', url, headers = headers, json=jsonParams) 
    
    if 500 <= resp.status_code <=600 and num_retries != 0:
        print(resp.status_code, resp.reason)
        return download(url, num_retries = num_retries-1)

    return resp

In [27]:
dj = download_json("http://httpbin.org/post")  
# post 방식 데이터 잘 받았는지 확인하는 주소

In [28]:
dj.text # \ -> escape 문자 잘 받았다는 뜻

'{"args":{},"data":"\\"{\\\\\\"where\\\\\\": \\\\\\"nexearch\\\\\\", \\\\\\"sm\\\\\\": \\\\\\"top_hty\\\\\\", \\\\\\"fbm\\\\\\": 1, \\\\\\"ie\\\\\\": \\\\\\"utf8\\\\\\", \\\\\\"query\\\\\\": \\\\\\"%ED%95%9C%EA%B5%AD\\\\\\"}\\"","files":{},"form":{},"headers":{"Accept":"*/*","Accept-Encoding":"gzip, deflate","Connection":"close","Content-Length":"113","Content-Type":"application/json","Host":"httpbin.org","User-Agent":"python bot"},"json":"{\\"where\\": \\"nexearch\\", \\"sm\\": \\"top_hty\\", \\"fbm\\": 1, \\"ie\\": \\"utf8\\", \\"query\\": \\"%ED%95%9C%EA%B5%AD\\"}","origin":"163.152.3.131","url":"http://httpbin.org/post"}\n'

4) requests.request - get 방식2

In [29]:
# get 방식
def download_get(url, agent="python bot", num_retries=2) :
    headers = {"user-agent" : agent} 
    resp = requests.get(url, headers = headers) 
    
    if 500 <= resp.status_code <=600 and num_retries != 0:
        print(resp.status_code, resp.reason)
        return download(url, num_retries = num_retries-1)

    return resp

In [30]:
dg = download_get(url, agent=agent)

In [31]:
dg.text

'<!doctype html> <html lang="ko"> <head> <meta charset="utf-8"> <meta name="referrer" content="always">  <meta name="format-detection" content="telephone=no,address=no,email=no"> <meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=2.0"> <meta property="og:title" content="한국 : 네이버 통합검색"/> <meta property="og:image" content="https://ssl.pstatic.net/sstatic/search/common/og_v3.png"> <meta property="og:description" content="\'한국\'의 네이버 통합검색 결과입니다."> <meta name="description" lang="ko" content="\'한국\'의 네이버 통합검색 결과입니다."> <title>한국 : 네이버 통합검색</title> <link rel="shortcut icon" href="https://ssl.pstatic.net/sstatic/search/favicon/favicon_140327.ico">  <link rel="search" type="application/opensearchdescription+xml" href="https://ssl.pstatic.net/sstatic/search/opensearch-description.https.xml" title="Naver" /><link rel="stylesheet" type="text/css" href="https://ssl.pstatic.net/sstatic/search/pc/css/search1_180712.css"> <link rel="stylesheet" type="text/css" href="https

5) requests.request - get 방식3

In [32]:
# get 방식 - url의 파라미터를 get의 알규먼트로 넣어주는 경우
def download_get_params(url, agent="python bot", num_retries=2) :
    headers = {"user-agent" : agent} 
    resp = requests.get(url, headers = headers, params=params) 
    
    if 500 <= resp.status_code <=600 and num_retries != 0:
        print(resp.status_code, resp.reason)
        return download(url, num_retries = num_retries-1)

    return resp

In [33]:
dgp = download_get_params("https://search.naver.com/search.naver", agent=agent) # 본주소만 적으면 된다

In [34]:
dgp.text # 한국어처리는 담에

'<!doctype html> <html lang="ko"> <head> <meta charset="utf-8"> <meta name="referrer" content="always">  <meta name="format-detection" content="telephone=no,address=no,email=no"> <meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=2.0"> <meta property="og:title" content="%ED%95%9C%EA%B5%AD : 네이버 통합검색"/> <meta property="og:image" content="https://ssl.pstatic.net/sstatic/search/common/og_v3.png"> <meta property="og:description" content="\'%ED%95%9C%EA%B5%AD\'의 네이버 통합검색 결과입니다."> <meta name="description" lang="ko" content="\'%ED%95%9C%EA%B5%AD\'의 네이버 통합검색 결과입니다."> <title>%ED%95%9C%EA%B5%AD : 네이버 통합검색</title> <link rel="shortcut icon" href="https://ssl.pstatic.net/sstatic/search/favicon/favicon_140327.ico">  <link rel="search" type="application/opensearchdescription+xml" href="https://ssl.pstatic.net/sstatic/search/opensearch-description.https.xml" title="Naver" /><link rel="stylesheet" type="text/css" href="https://ssl.pstatic.net/sstatic/search/pc/css/search1